### DISCUSSION TITLE
Latency?

### DISCUSSION THREAD

**Discussion by dhstsw** [probonopd/MiniDexed#345] [General]
Hi all,

my MiniDexed (Rpi2) is connected to a Raspberry PIco running the host midi adapter (https://github.com/rsta2/pico/tree/main/midi_adapter) and had zero problems so far.

Meaning, until i finally started tried to use it in a producttion evironement (meaning, connected to a DAW).

It works as expected and, while playing it live it's not noticeable, i found it to have quite a bit of latency compared to any other external midi device i have.

This is one bar of the classic BASS 1 (preset 15, bank 1) recorded on the beat:

![image](https://user-images.githubusercontent.com/59790194/188459505-7ff8bdc3-5522-4641-97e7-e68f79193b5e.png)

![image](https://user-images.githubusercontent.com/59790194/188459568-57ce291c-ae2c-421b-b023-07c5c9cf3a36.png)

Before you say, i know it's normal to have a bit of latency but this is really quite a lot.
Any idea if it may be due to the Pico?
Using a normal midi like the one below may be of any benefit?

![image](https://user-images.githubusercontent.com/59790194/188460087-b5f73c32-58c2-41e5-8541-6792e2fa90f3.png)

Thanks.

**Comment by stanleygarvey** [comment]
You don't say how you output audio, HDMI is a major cause of latency issues.

5 Sept 2022 14:24:12 dhstsw ***@***.***>:

**Reply by dhstsw** [reply]
PCM5102A

**Comment by stanleygarvey** [comment]
Same as mine, low latency. Have no experience with Pico I use usb midi.

5 Sept 2022 17:30:28 dhstsw ***@***.***>:

**Comment by stanleygarvey** [comment]
What DAW setup? Do you use Jackd if so check settings defaults of 1024 buffer too big, try lower 256.

5 Sept 2022 17:51:00 Stanley Garvey ***@***.***>:

**Reply by dhstsw** [reply]
DAW settings are the same i use with my other midi instruments without issues.
Anyway it's Ableton Live, ASIO buffer is set to 256.

**Comment by stanleygarvey** [comment]
Okay, your input and output looking good for low latency, can't help with Pico as I don't use one.

5 Sept 2022 18:44:39 dhstsw ***@***.***>:

**Comment by probonopd** [comment]
I'd like to do some latency testing. How do you measure latency? Please describe step-by-step, assuming no previous knowledge of the topic. Thanks!

**Comment by stanleygarvey** [comment]
Not scientific, just by ear. I play notes on piano and notes come out of mini dexed at the same time, Rpi3 and Rpi 2 v1and 2.
HDMI audio is very bad for latency as is large buffers on DAWs.
If you have little latency when you play live compared with higher latency when playing via DAW then the DAW seems to be the cause.

5 Sept 2022 19:56:10 probonopd ***@***.***>:

**Reply by dhstsw** [reply]
I think the problem is the serial midi to wich the pico is connected to: the pico is ways faster than needed to accomplish the work it has to do, but i don't have any way to prove if that's the correct assumption. I don't have one of those midi boards and you can't control it without a pico or midi board with a DAW.
Maybe in some months i'll source one and try again.
Hope to be able to sort this problem out: beside using track latency compensation in the daw (wich anyway has 'effect' only after the minidexed audio has been recorded) it would be (for me) a problem since the whole idea was to have it sequenced with a daw.

**Reply by dcoredump** [reply]
The best way to check the latency is to use a oscilloscope with two channels: one for showing the MIDI pulses and the other one for showing the audio answer. The time difference between the first rising of both signals is the latency.

At 44.100 Hz it should be around 2.9ms using a buffer of 128 bytes or 5.8ms with the default buffer of 256 bytes. 

I hope to manage to clean up my basement soon so I can set up the oscilloscope and take the measurement.
